### PR TITLE
migrate github.com/docker/docker/pkg/signal

### PR DESCRIPTION
**Pull request by thaJeztah**
relates to https://github.com/containerd/containerd/issues/5402

Migrating this package from commit:
https://github.com/docker/docker/commit/c81abefdb1f907bbc5f5b8b1b1fba942821ae5b3

Strategy taken:

    # install filter-repo (https://github.com/newren/git-filter-repo/blob/main/INSTALL.md)
    brew install git-filter-repo

    cd ~/projects

    # create a temporary clone of docker
    git clone https://github.com/docker/docker.git moby_signal
    cd moby_signal

    # remove all code, except for pkg/signal, and rename to /signal
    git filter-repo  --path pkg/signal --path-rename pkg/signal:signal

    # go to the target github.com/moby/sys repository
    cd ~/projects/moby-sys

    # create a branch to work with
    git checkout -b integrate_moby_signal

    # add the temporary repository as an upstream and make sure it's up-to-date
    git remote add moby_signal ~/projects/moby_signal
    git fetch moby_signal

    # merge the upstream code
    git merge --allow-unrelated-histories --signoff -S moby_signal/master

After this:

- signal: update import-paths
- signal: add go.mod
- Makefile: enable tests for signals
- signal: remove logrus dependency
- signal: remove gotest.tools dependency
- signal: remove github.com/pkg/errors dependency

### CODE REVIEW

File: signal/trap.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+						if atomic.AddUint32(&interruptCount, 1) == 1 {
+							// Call the provided cleanup handler
+							cleanup()
+							os.Exit(0)
+						} else {
+							return
+						}
+					} else {
+						// 3 SIGTERM/INT signals received; force exit without cleanup
+						logger.Info("Forcing docker daemon shutdown without cleanup; 3 interrupts received")
```
Comment (thaJeztah) [inline signal/trap.go:56]: This should probably be made more generic

```suggestion
						logger.Info("Forcing shutdown; 3 interrupts received")
```

File: signal/trap.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+					}
+				case syscall.SIGQUIT:
+					_, _ = DumpStacks("")
+					logger.Info("Forcing docker daemon shutdown without cleanup on SIGQUIT")
```
Comment (thaJeztah) [inline signal/trap.go:60]: Same for this:

```suggestion
					logger.Info("Forcing shutdown on SIGQUIT")
```

Comment (thaJeztah) [reply signal/trap.go:60]: Or perhaps this whole file should be kept in the moby repository (as it may be quite specific to moby)

**Comment by thaJeztah** [comment]
Some old commits were missing a proper DCO, so I manually marked it to "pass"

```
Commit sha: 6057481, Author: Guillaume J. Charmes, Committer: Guillaume J. Charmes; The sign-off is missing.
Commit sha: 108a12b, Author: Guillaume J. Charmes, Committer: Guillaume J. Charmes; The sign-off is missing.
Commit sha: aa57f84, Author: Guillaume J. Charmes, Committer: Guillaume J. Charmes; The sign-off is missing.
Commit sha: 5fcf5fa, Author: Kato Kazuyoshi, Committer: Kato Kazuyoshi; The sign-off is missing.
Commit sha: 530bcc0, Author: Guillaume J. Charmes, Committer: Guillaume J. Charmes; The sign-off is missing.
Commit sha: a3e2247, Author: Tianon Gravi, Committer: Tianon Gravi; The sign-off is missing.
Commit sha: 2f0e890, Author: Solomon Hykes, Committer: Solomon Hykes; The sign-off is missing.
Commit sha: 57a9d66, Author: Phil Estes, Committer: Phil Estes; The sign-off is missing.
Commit sha: 860f076, Author: Amit Krishnan, Committer: Amit Krishnan; Expected "Amit Krishnan amit.krishnan@oracle.com", but got "Amit Krishnan krish.amit@gmail.com".
```

**Comment by thaJeztah** [comment]
I guess this is one to decide (can be done later if we want)

> Or perhaps this whole file should be kept in the moby repository (as it may be quite specific to moby)

File: signal/trap.go (hunk 1 above)
Comment (AkihiroSuda) [reply signal/trap.go:56]: SGTM :+1:

**Comment by thaJeztah** [comment]
I'm updating the moby package in https://github.com/moby/moby/pull/42641 to get rid of the things we don't need

**Comment by thaJeztah** [comment]
closing in favour of (WIP) https://github.com/moby/sys/pull/70

**Pull request closed without merging**